core/src/main/scala/org/typelevel/otel4s/trace/Tracer.scala

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s
+package trace
+
+import cats.Applicative
+import cats.effect.kernel.Resource
+import org.typelevel.otel4s.meta.InstrumentMeta
+
+/** The API is responsible for creating [[Span]]'s.
+  */
+trait Tracer[F[_]] extends TracerMacro[F] {
+
+  /** Instrument metadata. Indicates whether instrumentation is enabled or not.
+    */
+  def meta: Tracer.Meta[F]
+
+  /** Returns trace identifier of a span that is available in a scope.
+    */
+  def traceId: F[Option[String]]
+
+  /** Returns span identifier of a span that is available in a scope.
+    */
+  def spanId: F[Option[String]]
```

Comment:
I realise this is WIP so not sure if you plan to change this or if the Otel spec has changed, but these are usually represented as byte arrays; e.g [TraceId](https://github.com/trace4cats/trace4cats/blob/master/modules/kernel/src/main/scala/trace4cats/model/TraceId.scala) and [SpanId](https://github.com/trace4cats/trace4cats/blob/master/modules/kernel/src/main/scala/trace4cats/model/SpanId.scala) in t4c. It would be good to be able to access these directly rather than just the string representation.

Apologies if I've jumped the gun and you're going to change these, or I've missed something and they're represented elsewhere!

Reply:
It's a draft version and my gut feeling says there will be a lot of changes :)
It would be interesting for me to hear your opinion about the current API.

Open Telemetry has the following info in span context:
```java
String getTraceId();
byte[] getTraceIdBytes();
String getSpanId();
byte[] getSpanIdBytes();
boolean isSampled();
TraceFlags getTraceFlags();
TraceState getTraceState();
boolean isValid();
boolean isRemote();
```

I believe we can incorporate the majority of methods.

Reply:
I'd be happy to see scodec-bits `ByteVector` used for byte arrays :) it's a stable dependency used in `fs2-core` and beyond.

Reply:
ah yeah, that makes sense. What we do in t4c is have `Show` instances for `TraceId` which makes those a bit more FP'y.

Also I made the the `isSampled` return [an ADT](https://github.com/trace4cats/trace4cats/blob/master/modules/kernel/src/main/scala/trace4cats/model/SampleDecision.scala) that can turn into a boolean, purely because everyone (including myself) kept getting confused as to what `true` meant when it came to sampling! 

In terms of the trace context in t4c it more closely follows the[ W3C standard](https://www.w3.org/TR/trace-context-1/), so in terms of structure probably doesn't align very well with the Otel spec.